### PR TITLE
feat: make icon delimiter modifiable in config

### PR DESCRIPTION
**Pull request by psimovec**
## Description
This PR makes the delimiter between icons modifiable in config file, instead of using forced space.
It can be modified by setting `delim="something"` in config.

Examples:
 `delim=" - "` 
image
`delim=" " #SPACE`
image
 `delim=" " # NARROW NO-BREAK SPACE`
image
Someone might prefer no delimiter, or possibly set background color/background color of the delimiter
`delim=" <span bgcolor='white'> </span> "`
image
(requires some patches for hyprland-rs and waybar-hyprland)

I am VERY open to feedback, it is my very first piece of code written in Rust and I have basically no knowledge about this programming language.

## Type of change

- [x] New feature (non-breaking change which adds functionality) - It should behave the same if the config is not modified
-  ~~Breaking change: If config file is not modified, spaces will disappear between icons.~~
- [x] set the default " ", when delim is not set in config
- [x] This change requires a documentation update 

## How Has This Been Tested?

No extensive testing, I have just tried modifying the delim option., (I should write at least 1 test)

## Future
In next PR/PRs I would like to make whole "workspace pattern" modifiable in config, specifically this:
`let text = format!("{id}: {apps}");`

And then the same for general "icon"/"icon with counter".. So it is possible to format it as people wish, for example setting a different color for the counter of icons.

### CODE REVIEW

**Comment by cyrinux** [comment]
Congrats @psimovec, very good start, I'm learning too with this project. I will fetch locally and test, maybe we can remove some clone().
Otherwise I'm agree for the idea of using formatter in more intentisive way too.

File: src/renamer/mod.rs [hunk 1]
```diff
@@ -14,6 +14,7 @@ mod macros;
 
 pub struct Renamer {
     workspaces: Mutex<FxHashSet<i32>>,
+    delim: String,
```
Comment (cyrinux) [inline src/renamer/mod.rs:17]:
```suggestion
```

File: src/renamer/mod.rs [hunk 2]
```diff
@@ -22,6 +23,7 @@ impl Renamer {
     pub fn new(cfg: Config, args: Args) -> Self {
         Renamer {
             workspaces: Mutex::new(FxHashSet::default()),
+            delim: cfg.config.delim.clone(),
```
Comment (cyrinux) [inline src/renamer/mod.rs:26]:
```suggestion
```

File: src/renamer/mod.rs [hunk 3]
```diff
@@ -87,8 +89,12 @@ impl Renamer {
                 .entry(workspace_id)
                 .or_insert_with(|| "".to_string());
 
+            let delim = match workspace.is_empty() {
+                false=>{self.cfg.lock()?.config.delim.clone()}
```
Comment (cyrinux) [inline src/renamer/mod.rs:93]:
```suggestion
                false=>self.cfg.lock()?.config.delim.to_string()
```

File: src/renamer/mod.rs [hunk 4]
```diff
@@ -87,8 +89,12 @@ impl Renamer {
...
+                true=>{"".to_string()}
```
Comment (cyrinux) [inline src/renamer/mod.rs:94]:
```suggestion
                true=>"".to_string()
```

**Comment by psimovec** [comment]
Thanks for the review, it looks better now.
(I'd squash commits + possibly rebase if needed when we decide the PR is ready to be merged, so i don't force push 5 times a day :D)

**Comment by psimovec** [comment]
@cyrinux do you prefer this PR as 1 commit or do you prefer 2 commits - one for code, one for docs?

File: src/renamer/mod.rs [hunk 5]
```diff
@@ -87,8 +87,12 @@ impl Renamer {
                 .entry(workspace_id)
                 .or_insert_with(|| "".to_string());
 
+            let delim = match workspace.is_empty() {
+                false=>self.cfg.lock()?.config.delim.to_string(),
+                true=>"".to_string()
+            };
             *workspace =
-                handle_new_icon(icon, client.fullscreen, workspace, should_dedup, *counter);
+                handle_new_icon(icon, delim , client.fullscreen, workspace, should_dedup, *counter);
```
Comment (cyrinux) [inline src/renamer/mod.rs:95]:
```suggestion
                handle_new_icon(icon, delim, client.fullscreen, workspace, should_dedup, *counter);
```

**Comment by psimovec** [comment]
> Please try to add the `[format]` config level and put the `delim` under, if you can't find how to do it, tell me, and I will do it after you. 🙏🏻

Should dedup be under the `[format]` too?

**Comment by cyrinux** [comment]
> > Please try to add the `[format]` config level and put the `delim` under, if you can't find how to do it, tell me, and I will do it after you. 🙏🏻
> 
> Should dedup be under the `[format]` too?

I was not sure about it but if you ask me, I feel like we should. Put it under format section as this all config will be under a section.

**Comment by cyrinux** [comment]
Another hint, if it is not already the code, use a code editor with the rust lsp enable this will help.

**Comment by psimovec** [comment]
I am struggling to initialize the FxHashMap to have default `delim=" "`, I commented it out for now
I'll try to install the  extension you mentioned

**Comment by psimovec** [comment]
Alright even chatgpt3 is way better programming than me, it solved the problem with default delim for me, code doesn't look that nice but works :D

File: src/config/mod.rs [hunk 6]
```diff
@@ -19,14 +20,25 @@ pub struct ConfigFileRaw {
...
     #[serde(default)]
     pub dedup: bool,
 }
 
+
+fn default_format() -> FxHashMap<String, String> {
+    let mut map = FxHashMap::default();
+    map.insert("delim".to_string(), " ".to_string());
+    map
+}
```
Comment (cyrinux) [inline src/config/mod.rs:34]:
```suggestion
fn default_format() -> FxHashMap<String, String> {
    [("delim".to_string(), " ".to_string())].iter().cloned().collect()
}
```

**Comment by cyrinux** [comment]
> Alright even chatgpt3 is way better programming than me, it solved the problem with default delim for me, code doesn't look that nice but it seems to work :D

Let me use chatgpt4 to try to remove the mut 😄

**Comment by psimovec** [comment]
> 
> Let me use chatgpt4 to try to remove the mut smile

I guess time to buy chatGPT subcription again :D

**Comment by cyrinux** [comment]
> I am struggling to initialize the FxHashMap to have default `delim=" "`, I commented it out for now I'll try to install the extension you mentioned

We can try to do move dedup under format

**Comment by psimovec** [comment]
I kinda did it, but I don't like the way I did it, right now you would have to use something like `dedup="true"` which is really weird.. 
Not sure if I will have more time today+tomorrow, so I am totally fine if you modify this PR as you wish/add your changes so it looks/works better

**Comment by cyrinux** [comment]
Ok I will try to progress on this top of your commit. It's already a good progress!

**Comment by cyrinux** [comment]
@psimovec if you want I try to do it from my side then I guide you when you can. If you really want I make it and you will read then, tell me.

**Comment by cyrinux** [comment]
> > @cyrinux do you prefer this PR as 1 commit or do you prefer 2 commits - one for code, one for docs?
> 
> As you wish, but don't worry with number of commits, we can rebase at the end before get merge to include the feature. I would prefer merge add the end one commit with the whole stuff for the PR.

My co-maintainer @maximbaz prefer we dont squash or rebase, so feel free to just commit each of your steps for the future 😉

**Comment by cyrinux** [comment]
@psimovec I merge this PR that superseed this one, I make my work on top of yours  https://github.com/hyprland-community/hyprland-autoname-workspaces/pull/49
You can try the formatters in the master branch now ! 😄

**Pull request closed without merging**